### PR TITLE
Refactor Dockerfile to multi-staged docker builds

### PR DESCRIPTION
**Pull request by DanielMSchmidt**
Hi there,

First of all, thanks for this great project 👍 

Docker now supports [multi-staged builds](https://docs.docker.com/engine/userguide/eng-image/multistage-build/) which allows us not use two images manually but have all the build process in one image. I thought this might help, if you disagree I won't take offence 🤷‍♂️

### CODE REVIEW

**Comment by piosz** [comment]
@luxas could you please take a look?

**Comment by DirectXMan12** [comment]
we'd have to make sure whatever is running the builds supports a sufficiently recent version of Docker

File: influxdb/Dockerfile [hunk 1]
```diff
@@ -1,6 +1,14 @@
+FROM golang:GOLANG_VERSION as builder
+WORKDIR /go/src/github.com/influxdb/influxdb
+RUN curl -sSL https://github.com/influxdata/influxdb/archive/VERSION.tar.gz | tar -xz --strip-components=1
+RUN go get github.com/sparrc/gdm 
+RUN gdm restore 
+RUN ln -s /go/src/github.com/influxdb/influxdb /go/src/github.com/influxdata/ 
+RUN GOamd64=ARCH CGO_ENABLED=0 go build -a --installsuffix cgo --ldflags='-s' -o /build/influxd ./cmd/influxd
```
Comment (luxas) [inline influxdb/Dockerfile]: GOARCH=BUILDARCH?

File: influxdb/Makefile [hunk 2]
```diff
@@ -49,15 +49,14 @@ build:
...
-		&& go get github.com/sparrc/gdm \
-		&& gdm restore \
-		&& ln -s /go/src/github.com/influxdb/influxdb /go/src/github.com/influxdata/ \
-		&& GOARCH=$(ARCH) CGO_ENABLED=0 go build -a --installsuffix cgo --ldflags=\"-s\" -o /build/influxd ./cmd/influxd"
+	cd $(TEMP_DIR) && \
+	echo $(TEMP_DIR) && \
+	sed -i- "s|BASEIMAGE|$(BASEIMAGE)|g" Dockerfile && \
+	sed -i- "s|GOLANG_VERSION|$(GOLANG_VERSION)|g" Dockerfile && \
+	sed -i- "s|VERSION|$(VERSION)|g" Dockerfile && \
+	sed -i- "s|ARCH|$(ARCH)|g" Dockerfile && \
```
Comment (luxas) [inline influxdb/Makefile]: `s|BUILDARCH|$(ARCH)|g`

**Review by luxas — COMMENTED** [review]
I guess this would be fine, but I'm not sure we want to put a dep of a high docker version for everyone running this script as @DirectXMan12 said, what do others think?

**Comment by DanielMSchmidt** [comment]
You are right, forcing a new docker version on everyone is probably a bad idea 🤔

**Pull request closed without merging**